runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -12121,7 +12205,63 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
 
          if ((node->isUnsafeGetPutCASCallOnNonArray() || !TR::Compiler->om.canGenerateArraylets()) && node->isSafeForCGToFastPathUnsafeCall())
             {
-            resultReg = VMinlineCompareAndSwapObject(node, cg);
+            resultReg = VMinlineCompareAndSetOrExchangeReference(node, cg, false);
+            return true;
+            }
+         break;
+
+      case TR::jdk_internal_misc_Unsafe_compareAndExchangeInt:
+         // As above, we only want to inline the JNI methods, so add an explicit test for isNative()
+         if (!methodSymbol->isNative())
```

Comment:
The original problem traces back to this commit: https://github.com/eclipse-openj9/openj9/commit/111f0b0bcc55e3ae7d88900d0dbcca460e7ed875

Going from Java 8 to 11, `compareAndSwap[Object|Int|Long]` were changed to being wrappers to call Native versions with the same name. So there is both a java and native that matches the same recognized method.

Going from 11 to 17, `compareAndSetObject` and `compareAndExchangeObject` were changed from native to Java wrapper that calls the new names `compareAndSetReference` and `compareAndExchangeReference` that do the same thing.

That commit I linked to deals with `compareAndSwap[Object|Int|Long]` and `compareAndSetObject`. Acceleration for `compareAndExchangeObject` didn't exist so that case didn't need to be handled.

Based on the comment in the commit, the change explicitly went in to deal with incorrect behavior. But, it doesn't say what it is and I'm not exactly sure what the problem is either.

I copied the same check for the acceleration of `compareAndExchange[Reference|Int|Long]`. But, those are always native in every JDK level they exist in. So I can remove the check for those cases.

`compareAndSet[Reference|Int|Long]` is also always native but uses the same recognized methods as `compareAndSwap[Object|Int|Long]` so I can't remove the check for that case without updating Z and arm code which is currently future work.

Reply:
I talked to @JamesKingdon and he pointed me towards this issue from around 7 years ago: https://github.com/eclipse-openj9/openj9/issues/321

From the issue the test would hang. The exact reasoning is not listed but somehow it was determined that preventing the non-native wrapped calls of the CAS methods from being turned into accelerated assembly made the problem stop. As a conservative fix, the optimization was limited to just the native methods and to leave it to the inliner to inline the wrapper methods.

Reply:
The linked issue was create due to an external test that I do not have access to. The issue also includes a simple program which could reproduce the issue but it no longer runs. The is mention in the issue of a dependency on a server that I assume is no longer running. So it does not seem to be possible to reproduce the problem.

I suspect that the original fix was only hiding an issue because it doesn't make much sense that it would have actually fixed the problem. Putting down accelerated assembly should have worked when replacing non-native Unsafe CAS call wrappers. As it would turn out, while creating this PR I encountered an issue with reference counting inside the X fast path generation for Unsafe CAS. While a different path is taken today, 7 years ago `compareAndSwapObject`, which appeared to cause the problem, went through `inlineCompareAndSwapNative` and that's where I found the problem with reference counting. It is also possible the problem was something else fixed in the last 7 years. Without being able to run the test, I can't confirm that.